src/operator/distinct.rs

```diff
@@ -0,0 +1,288 @@
+                self.circuit()
+                    .add_unary_operator(UnaryOperatorAdapter::new(Distinct::new()), self)
+            })
+            .clone()
+    }
+
+    /// Incremental version of the [`Distinct`] operator.
+    ///
+    /// This is equivalent to `self.integrate().distinct().differentiate()`, but
+    /// is more efficient.
+    pub fn distinct_incremental<V, W>(&self) -> Stream<Circuit<P>, <Z as SharedRef>::Target>
+    where
+        V: KeyProperties,
+        W: ZRingValue,
+        Z: SharedRef + 'static,
+        <Z as SharedRef>::Target: ZSet<V, W> + SharedRef<Target = Z::Target>,
+        for<'a> &'a <Z as SharedRef>::Target: IntoIterator<Item = (&'a V, &'a W)>,
+    {
+        self.circuit()
+            .cache_get_or_insert_with(DistinctIncrementalId::new(self.local_node_id()), || {
+                self.circuit().add_binary_operator(
+                    BinaryOperatorAdapter::new(DistinctIncremental::new()),
+                    self,
+                    &self.integrate().delay(),
+                )
+            })
+            .clone()
+    }
+
+    /// Incremental nested version of the [`Distinct`] operator.
+    pub fn distinct_incremental_nested<V, W>(&self) -> Stream<Circuit<P>, <Z as SharedRef>::Target>
+    where
+        V: KeyProperties,
+        W: ZRingValue,
+        Z: SharedRef + 'static,
+        <Z as SharedRef>::Target: ZSet<V, W> + SharedRef<Target = Z::Target>,
+        for<'a> &'a <Z as SharedRef>::Target: IntoIterator<Item = (&'a V, &'a W)>,
+    {
+        self.integrate_nested()
+            .distinct_incremental()
+            .differentiate_nested()
+    }
+}
+
+/// Distinct operator changes all weights in the support of a Z-set to 1.
+pub struct Distinct<V, W, Z> {
+    _type: PhantomData<(V, W, Z)>,
+}
+
+impl<V, W, Z> Distinct<V, W, Z> {
+    pub fn new() -> Self {
+        Self { _type: PhantomData }
+    }
+}
+
+impl<V, W, Z> Default for Distinct<V, W, Z> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<V, W, Z> Operator for Distinct<V, W, Z>
+where
+    V: 'static,
+    W: 'static,
+    Z: 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("Distinct")
+    }
+    fn clock_start(&mut self, _scope: Scope) {}
+    fn clock_end(&mut self, _scope: Scope) {}
+}
+
+impl<V, W, Z> UnaryOperator<Z, Z> for Distinct<V, W, Z>
+where
+    V: KeyProperties,
+    W: ZRingValue,
+    Z: ZSet<V, W> + 'static,
+{
+    fn eval(&mut self, i: &Z) -> Z {
+        i.distinct()
+    }
+
+    fn eval_owned(&mut self, i: Z) -> Z {
+        i.distinct_owned()
+    }
+}
+
+/// Incremental version of the distinct operator.
+///
+/// Takes a stream `a` of changes to relation `A` and a stream with delayed
+/// value of `A`: `z^-1(A) = a.integrate().delay()` and computes
+/// `distinct(A) - distinct(z^-1(A))` incrementally, by only considering
+/// values in the support of `a`.
+struct DistinctIncremental<V, W, Z> {
```

Comment:
Technically this is LiftedDistinctIncremental.

Reply:
I have so far omitted the `Lifted` part in all operator names.  It somehow doesn't seem to make anything clearer.

Reply:
Because for most it doesn't make a difference